python/ql/src/experimental/Security/CWE-352/DjangoCSRF.ql

```diff
@@ -0,0 +1,36 @@
+/**
+ * @name Disabled Django CSRF protection
+ * @description No csrf Middleware/ Decorator to protect Django application from CSRF
+ * @kind problem
+ * @problem.severity error
+ * @precision high
+ * @id py/django-csrf
+ * @tags security
+ *       external/cwe/cwe-352
+ */
+
+import python
+import semmle.python.strings
+
+from GlobalVariable g
+where
+  /*
+   * Regex matching MIDDLEWARE & MIDDLEWARE_CLASSES(backwards compatibility)
+   * Checks if there is a GlobalVariable(List/Tuple) named MIDDLEWARE or MIDDLEWARE_CLASES
+   * and contains the middleware string
+   */
+
+  g.getId().regexpMatch("MIDDLEWARE(_CLASSES)?") and
+  not exists(List l |
+    l.getParentNode().getAChildNode().toString().regexpMatch(g.getId()) and
+    exists(StrConst mw | mw.getS().regexpMatch(".*csrf.*") and l.contains(mw))
+  ) and
+  not exists(Tuple t |
+    t.getParentNode().getAChildNode().toString().regexpMatch(g.getId()) and
+    exists(StrConst mw | mw.getS().regexpMatch(".*csrf.*") and t.contains(mw))
+  ) and
+  /* Check if any CSRF decorator is used (requires_csrf_token | ensure_csrf_cookie  | csrf_protect) */
+  not exists(FunctionExpr f |
+    f.getADecoratorCall().getFunc().toString().regexpMatch(".*csrf_(protect|exempt|cookie).*")
```

Comment:
After reading through the [django documentation](https://docs.djangoproject.com/en/3.0/ref/csrf/#module-django.views.decorators.csrf), it is only `django.views.decorators.csrf.csrf_protect` that offers any CSRF protection. 

As described [here](https://docs.djangoproject.com/en/3.0/ref/csrf/#django.views.decorators.csrf.requires_csrf_token) `requires_csrf_token` and `ensure_csrf_cookie` only helps in _adding a CSRF token to the response_.

Instead of using `toString()`, we should look at what the value of the decorator is, like so:

```suggestion
    f.getADecorator().pointsTo() = Value::named("django.views.decorators.csrf.csrf_protect")
```